src/Rings/mpoly-graded.jl

```diff
@@ -1768,65 +1767,66 @@ function _hilbert_numerator_from_leading_exponents(
     # short exponent vectors where the k-th bit indicates that the k-th 
     # exponent is non-zero.
   )
-  ret = _hilbert_numerator_trivial_cases(a, weight_matrix, return_ring)
+  t = gens(return_ring)
```

Comment:
Is this to avoid the allocation of the vector `t` within all the function calls? I would not have guessed that this is really an issue since the recursion depth is not that deep, usually.

Reply:
It actually is; computing `gens(S)` each time allocates new polynomials, and a `Vector` to contain them. And polynomials also require multiple allocation each.

This is particular bad if you write a loop with `gens(S)[i]` ! For this, `gen(S, i)` is much better. But even better is to do what I did here, and compute `gens(S)` only once at the start and then reuse it.

Best would be (IMHO) if `gens(S)` was an attribute, i.e., cached; but this would be risky; we need "ImmutableVector" for that, and ideally also a flag in the polynomials to ensure they are not modified... So right now it's not practical.